horovod/spark/common/util.py

```diff
@@ -0,0 +1,518 @@
+from pyspark.ml.linalg import DenseVector, SparseVector, VectorUDT
+from pyspark.sql.types import (IntegerType, StringType, FloatType,
+                               BinaryType, DoubleType, LongType, BooleanType,
+                               ArrayType)
+from pyspark.sql.types import from_arrow_type
+
+from horovod.spark.common import cache, constants
+
+_training_cache = cache.TrainingDataCache()
+
+
+def data_type_to_str(dtype):
+    if dtype == VectorUDT:
+        return 'Vector'
+    elif dtype == IntegerType:
+        return 'Int'
+    elif dtype == StringType:
+        return 'String'
+    elif dtype == FloatType:
+        return 'Float'
+    elif dtype == BinaryType:
+        return 'Binary'
+    elif dtype == DoubleType:
+        return 'Double'
+    elif dtype == LongType:
+        return 'Long'
+    elif dtype == BooleanType:
+        return 'Boolean'
+    else:
+        raise ValueError('Unrecognized data type: {}'.format(dtype))
+
+
+def numpy_type_to_str(dtype):
+    if dtype == np.int32:
+        return 'Int'
+    elif dtype == np.float32:
+        return 'Float'
+    elif dtype == np.uint8:
+        return 'Binary'
+    elif dtype == np.float64:
+        return 'Double'
+    elif dtype == np.int64:
+        return 'Long'
+    elif dtype == np.bool:
+        return 'Boolean'
+    else:
+        raise ValueError('Cannot convert numpy data type to Spark string: {}'.format(dtype))
+
+
+def spark_scalar_to_python_type(dtype):
+    if dtype == IntegerType:
+        return int
+    elif dtype == StringType:
+        return str
+    elif dtype == FloatType:
+        return float
+    elif dtype == DoubleType:
+        return float
+    elif dtype == LongType:
+        return int
+    elif dtype == BooleanType:
+        return bool
+    elif dtype == BinaryType:
+        return bytes
+    else:
+        raise ValueError('cannot convert Spark data Type {} to native python type'.format(dtype))
+
+
+def pyarrow_to_spark_data_type(dtype):
+    # PySpark will interpret list types as Arrays, but for ML applications we want to default to
+    # treating these as DenseVectors.
+    if pa.types.is_list(dtype):
+        return DenseVector
+    return type(from_arrow_type(dtype))
+
+
+def data_type_to_numpy(dtype):
+    if dtype == VectorUDT or dtype == SparseVector or dtype == DenseVector:
+        return np.float64
+    elif dtype == ArrayType:
+        return np.float64
+    elif dtype == IntegerType:
+        return np.int32
+    elif dtype == StringType:
+        return np.uint8
+    elif dtype == FloatType:
+        return np.float32
+    elif dtype == BinaryType:
+        return np.uint8
+    elif dtype == DoubleType:
+        return np.float64
+    elif dtype == LongType:
+        return np.int64
+    elif dtype == BooleanType:
+        return np.bool
+    else:
+        raise ValueError('Unrecognized data type: {}'.format(dtype))
+
+
+def check_shape_compatibility(metadata, feature_columns, label_columns,
```

Comment:
all errors raised here should be tested on unit tests

Reply:
Done.